### PR TITLE
feat: initial version to support sampling and drop rules api and deployment

### PR DESCRIPTION
**Pull request by mindhash**
This PR:
- introduces APIs for sampling and drop rules
- config versioning and deployment to agent (collector configs)
- some changes to opamp for building processor config from sampling/drop rules

Related to: https://github.com/SigNoz/engineering-pod/issues/935

Pending:
- multi-agent support through opamp

### CODE REVIEW

File: pkg/query-service/app/opamp/pipeline_builder.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+				if lastMatched <= 0 {
+					zap.S().Debugf("build_pipeline: found a new item to be inserted, inserting at position 0:", m.Name)
+					pipeline = append([]interface{}{m.Name}, pipeline[lastMatched+1:]...)
+				} else {
+					zap.S().Debugf("build_pipeline: found a new item to be inserted, inserting at position :", lastMatched, " ", m.Name)
+					prior := pipeline[:lastMatched]
+					next := pipeline[lastMatched+1:]
+
+					pipeline = append(prior, []interface{}{m.Name})
```
Comment (nityanandagohain) [inline pkg/query-service/app/opamp/pipeline_builder.go:180]: It should be `pipeline = append(prior, m.Name)`

File: pkg/query-service/app/opamp/pipeline_builder.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+				// we use last matched to insert new item.  This means, we keep inserting missing processors
+				// right after last matched processsor (e.g. insert filters after tail_sampling for existing list of [batch, tail_sampling])
+
+				if lastMatched <= 0 {
+					zap.S().Debugf("build_pipeline: found a new item to be inserted, inserting at position 0:", m.Name)
+					pipeline = append([]interface{}{m.Name}, pipeline[lastMatched+1:]...)
+				} else {
+					zap.S().Debugf("build_pipeline: found a new item to be inserted, inserting at position :", lastMatched, " ", m.Name)
+					prior := pipeline[:lastMatched]
+					next := pipeline[lastMatched+1:]
```
Comment (nityanandagohain) [inline pkg/query-service/app/opamp/pipeline_builder.go:178]: Not the correct way to copy `next` as when pipeline will be modified in the next line, the value of next will change which will result in wrong result.


```
prior := make([]interface{}, len(pipeline[:lastMatched]))
next := make([]interface{}, len(pipeline[lastMatched:]))
copy(prior, pipeline[:lastMatched])
copy(next, pipeline[lastMatched:])

pipeline = append(prior, m)
pipeline = append(pipeline, next...)
```

File: ee/query-service/ingestionRules/db.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	if err != nil {
+		return nil, []error{fmt.Errorf("failed to get drop rules from db: %v", err)}
+	}
+
+	if len(rules) == 0 {
+		return rules, nil
+	}
+
+	for _, d := range rules {
+		if err := d.ParseRawConfig(); err != nil {
```
Comment (nityanandagohain) [inline ee/query-service/ingestionRules/db.go:120]: This won't work as it is not a pointer https://forum.golangbridge.org/t/update-values-in-a-struct-through-a-method/19589

Comment (mindhash) [reply ee/query-service/ingestionRules/db.go:120]: I will test this out. I think it was working. My understanding is slice stores a point to struct. so should work. but thanks for pointing out. I will confirm

**Pull request closed without merging**